tensorflow/python/framework/device_spec.py

```diff
@@ -333,17 +333,21 @@ def _string_to_components(spec=None):
           replica = y[1]
         elif ly == 2 and y[0] == "task":
           task = y[1]
-        elif ((ly == 1 or ly == 2) and (y[0].upper() in _VALID_DEVICE_TYPES)):
+        elif ((ly == 1 or ly == 2) and (y[0] != "")):
```

Comment:
We shouldn't remove this check hust because PluggableDevices can have any names. Please create a new list of valid device types consisting of `_VALID_DEVICE_TYPES` and the device types found in `tf.config.list_physical_devices()`, and check if `y[0]` is in this new list instead.

Reply:
I think it may impact the user experience if the device type check is depend on the installed device. 
1. If user installed a plugin which register an XPU device, but not install the hardware yet. user may expect the following code will fallback to cpu since `allow_soft_placement` is true. If with list_physical_devices check, this will report invalid device name
    ```
    config=tf.ConfigProto(allow_soft_placement=True, log_device_placement=True)
   with tf.device("XPU:0")
       xxx
    ```
2. user installed the XPU plugin on two machine, one has XPU device and another don't have.  one is running with fine, another report XPU is not a valid device name, I think this may confusing? thanks.
   ```
   with tf.device("XPU:0")
   xxx
   ```

Reply:
True. It's possible that people could have plug-in without the device. What about getting the types from all registered devices instead? Could there be potential problems with this?

Reply:
I talked to @allenlavoie and he prefers still being able to check for typos (e.g., in case users type GPUU instead of GPU), but he is also okay with removing the check if that's not too hard. 

Should we move `device_spec.py` and `device_test.py` to a new PR and discuss there so this won't block the main implementation?

Reply:
Yeah, my worry is that soft placement will "allow" these typos ("GPUU", "INVALID", etc. implicitly mapping to "CPU") and backward compatibility will make it hard to raise an error again. We already have this somewhat-bad soft placement for "GPU:999" (implicitly mapping to "GPU:0").

So if we can generate a list of plugin devices and just add them to the allowed list that'd be ideal.

Reply:
@penpornk @allenlavoie  I have modified the code and make (PhysicalDevice types | VALID_DEVICE_TYPES) as the valid device types set. please have a review. Thanks